libraries/Microsoft.Bot.Builder.Skills/ChannelApiMiddleware.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Skills
+{
+    /// <summary>
+    /// Handles InvokeActivity for ChannelAPI methods calls coming from SkillHostController.
+    /// </summary>
+    internal class ChannelApiMiddleware : IMiddleware
+    {
+        private readonly SkillAdapter _skillAdapter;
+
+        internal ChannelApiMiddleware(SkillAdapter skillAdapter)
+        {
+            _skillAdapter = skillAdapter;
+        }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
+        {
+            // register the skill adapter so people can get it to do .ForwardActivityAsync()
+            turnContext.TurnState.Add(_skillAdapter);
+
+            if (turnContext.Activity.Type == ActivityTypes.Invoke && turnContext.Activity.Name == "ChannelAPI")
+            {
+                // process Invoke Activity 
+                var invokeActivity = turnContext.Activity.AsInvokeActivity();
+                var invokeArgs = invokeActivity.Value as ChannelApiArgs;
+
+                // TODO This needs to be more robust to get bot id
+                await CallChannelApiAsync(turnContext, next, invokeArgs, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // just pass through
+                await next(cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task CallChannelApiAsync(ITurnContext turnContext, NextDelegate next, ChannelApiArgs invokeArgs, CancellationToken cancellationToken)
+        {
+            switch (invokeArgs.Method)
+            {
+                // Send activity(activity)
+                case ChannelApiMethods.SendToConversation:
+                    {
+                        var activityPayload = (Activity)invokeArgs.Args[0];
+                        if (activityPayload.Type == ActivityTypes.EndOfConversation)
```

Comment:
>if (activityPayload.Type == ActivityTypes.EndOfConversation [](start = 24, length = 59)

we want to pass through events as well. not just endofconversation.

Reply:
Hi Steve, so channels do not handle event activities? if yes, how do you propose we distinguish which ones go to the bot and which ones get posted back to the channel?